src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java

```diff
@@ -103,12 +103,14 @@ public void success(Object returnValue) {
 
     public static void success(String k, Object returnValue) {
         State s = State.get();
-        if (s.contexts.containsKey(k)) {
-            System.err.println("Unblocking " + k + " as success");
-            getContext(s, k).onSuccess(returnValue);
-        } else {
-            System.err.println("Planning to unblock " + k + " as success");
-            s.returnValues.put(k, returnValue);
+        synchronized (s) {
+            if (s.contexts.containsKey(k)) {
+                System.err.println("Unblocking " + k + " as success");
+                getContext(s, k).onSuccess(returnValue);
```

Comment:
I am somewhat concerned about holding a lock while calling `onSuccess` here. Is it effective to hold the lock on `State` only while accessing its fields? For example

```java
public static void success(String k, Object returnValue) {
    State s = State.get();
    synchronized (s) {
        if (!s.contexts.containsKey(k)) {
            System.err.println("Planning to unblock " + k + " as success");
            s.returnValues.put(k, returnValue);
            return;
        }
    }
    System.err.println("Unblocking " + k + " as success");
    getContext(s, k).onSuccess(returnValue);
}
```

Reply:
Yeah that's even better.